### PR TITLE
Add basic AnyDTF FFI

### PR DESCRIPTION
**Pull request by Manishearth**
sufficient for demos

I'm not happy with `try_new_from_iso_in_calendar`, but I'm also worried about having a `try_new_from_iso()` that keeps allocating an `Arc::new(AnyCalendar::Iso)`

I want to land this for demo usage, and we can come up with good FFI APIs next week :)

### CODE REVIEW

**Comment by sffc** [comment]
Please don't merge this until after I've released all the crates

**Comment by Manishearth** [comment]
I've updated the wasm demo, @QnnOkabayashi can you have a look as well? I just changed the existing demo to use calendars. It works pretty well!

@sffc sounds good; in the meantime I can use the PR preview artifact link if I need.

**Comment by Manishearth** [comment]
```
at/wasm/icu4x/lib/ICU4XBidiInfo.d.ts b/ffi/diplomat/wasm/icu4x/lib/ICU4XBidiInfo.d.ts
index 20afcef..950627c 100644
--- a/ffi/diplomat/wasm/icu4x/lib/ICU4XBidiInfo.d.ts
+++ b/ffi/diplomat/wasm/icu4x/lib/ICU4XBidiInfo.d.ts
@@ -1,4 +1,4 @@
-import { u8, u32, usize } from "./diplomat-runtime"
+import { u8, u32 } from "./diplomat-runtime"
 import { ICU4XBidiParagraph } from "./ICU4XBidiParagraph";
 
 /**
diff --git a/ffi/diplomat/wasm/icu4x/lib/ICU4XBidiParagraph.d.ts b/ffi/diplomat/wasm/icu4x/lib/ICU4XBidiParagraph.d.ts
index 7c4eabf..1d0cc61 100644
--- a/ffi/diplomat/wasm/icu4x/lib/ICU4XBidiParagraph.d.ts
+++ b/ffi/diplomat/wasm/icu4x/lib/ICU4XBidiParagraph.d.ts
@@ -1,4 +1,4 @@
-import { u8, u32, usize } from "./diplomat-runtime"
+import { u8, u32 } from "./diplomat-runtime"
 import { FFIError } from "./diplomat-runtime"
 import { ICU4XBidiDirection } from "./ICU4XBidiDirection";
 import { ICU4XError } from "./ICU4XError";
diff --git a/ffi/diplomat/wasm/icu4x/lib/ICU4XPluralOperands.d.ts b/ffi/diplomat/wasm/icu4x/lib/ICU4XPluralOperands.d.ts
index 5ab7574..aa9d025 100644
--- a/ffi/diplomat/wasm/icu4x/lib/ICU4XPluralOperands.d.ts
+++ b/ffi/diplomat/wasm/icu4x/lib/ICU4XPluralOperands.d.ts
@@ -1,4 +1,4 @@
-import { u64, u32, usize } from "./diplomat-runtime"
+import { u64, u32 } from "./diplomat-runtime"
 import { FFIError } from "./diplomat-runtime"
 import { ICU4XError } from "./ICU4XError";
 
diff --git a/ffi/diplomat/wasm/icu4x/lib/diplomat-runtime.d.ts b/ffi/diplomat/wasm/icu4x/lib/diplomat-runtime.d.ts
index f87bf5e..a6d8597 100644
--- a/ffi/diplomat/wasm/icu4x/lib/diplomat-runtime.d.ts
+++ b/ffi/diplomat/wasm/icu4x/lib/diplomat-runtime.d.ts
@@ -16,4 +16,3 @@ export type i64 = bigint;
 export type f32 = number;
 export type f64 = number;
 export type char = string;
-export type usize = number;
```

**Comment by Manishearth** [comment]
(working on fixing the error, probably need a merge from upstream)

**Review by sffc — DISMISSED** [review]
The "regen" commit regresses on the fix I had to make involving `usize` in order to get `tsc` to compile.

**Comment by QnnOkabayashi** [comment]
> The "regen" commit regresses on the fix I had to make involving `usize` in order to get `tsc` to compile.

I can make a PR on diplomat for this now

**Comment by Manishearth** [comment]
Ah, sounds like we need a diplomat bump, I had assumed that had already happened

**Comment by QnnOkabayashi** [comment]
https://github.com/rust-diplomat/diplomat/pull/219 should fix Shane's issue

File: ffi/diplomat/src/calendar.rs
Comment (sffc) [inline ffi/diplomat/src/calendar.rs:46]: Nit: Not `ICU4XGregorianDateTime`

File: ffi/diplomat/src/calendar.rs
Comment (sffc) [inline ffi/diplomat/src/calendar.rs:58]: Nit: This is an important section of code, and it's not clear what it is doing (selecting an extension keyword from a locale). I think I'd prefer if we didn't `impl TryFrom<Locale> for AnyCalendarKind`

File: ffi/diplomat/src/calendar.rs
Comment (sffc) [inline ffi/diplomat/src/calendar.rs:43]: Question: why `ICU4XCalendar` and not `ICU4XAnyCalendar`? How do you anticipate us representing typed calendars and dates in FFI?

File: ffi/diplomat/src/calendar.rs
Comment (sffc) [inline ffi/diplomat/src/calendar.rs:68]: Nit: Not Gregorian

File: ffi/diplomat/src/calendar.rs
Comment (sffc) [inline ffi/diplomat/src/calendar.rs:82]: Thought: I've said this before, and I'll say it again; I think we should allow for fractional seconds. We support fractional seconds in code and patterns, and there are real use cases for them, and we shouldn't demote them to some future second-class constructor. I acknowledge that we didn't end up doing that in the Rust API.

File: ffi/diplomat/src/datetime.rs
Comment (sffc) [inline ffi/diplomat/src/datetime.rs:193]: Nit: Fix docs

File: ffi/diplomat/src/datetime.rs
Comment (sffc) [inline ffi/diplomat/src/datetime.rs:198]: Nit: Fix docs

**Review by sffc — CHANGES_REQUESTED** [review]
Started actually reviewing the PR. (I hadn't actually done that before)

File: ffi/diplomat/src/calendar.rs
Comment (Manishearth) [reply ffi/diplomat/src/calendar.rs:43]: ICU4XJapaneseCalendar etc

"Any" is a Rusty term and only really makes sense to me when there's a generic to compare against, which we don't have here.

Also I'd imagine that we won't actually have individual calendars exposed over FFI and just use AnyCalendar for everything. It's far nicer over FFI and it follows our history of using dynamic dispatch to deal with generics over FFI.

File: ffi/diplomat/src/calendar.rs
Comment (Manishearth) [reply ffi/diplomat/src/calendar.rs:82]: To be clear, the intent is for them to be mutated in, not in a constructor. I can add a mutation API, I was trying to add the minimal APIs for the demo.

I think they would be fine in a constructor that took in a struct, but otherwise with just ordered parameters it's unclear what the last parameter is from context (and it makes the identity of the other parameters confusing), which is why I've been reluctant to have this in the ctor without having `with_fractional_seconds` in its name

File: ffi/diplomat/src/calendar.rs
Comment (Manishearth) [reply ffi/diplomat/src/calendar.rs:43]: (I don't have a strong opinion here though)

File: ffi/diplomat/src/calendar.rs
Comment (Manishearth) [reply ffi/diplomat/src/calendar.rs:82]: Also I think in our final version we should also have a from_codes ctor, I just didn't add it here because I was going for minimal necessary for the demo

**Comment by Manishearth** [comment]
```
Error handling response: TypeError: Cannot read properties of undefined (reading 'filter')
```

when attempting to use the calendar selector

**Comment by Manishearth** [comment]
Hmmm, the calendar selector seems _more_ broken now in https://storage.googleapis.com/icu4x-pr-artifacts/gha/4bde4e3ac3c6600fe7b24692863b669181c115fa/wasm-demo/index.html

but "use from locale" is still working

**Comment by Manishearth** [comment]
Bug fixed

File: ffi/diplomat/src/calendar.rs
Comment (Manishearth) [reply ffi/diplomat/src/calendar.rs:58]: I kinda see this as analogous to the DataLocale inputs for most types; AnyCalendar directly takes an AnyCalendarKind but that's really not that different from taking in a DataLocale and reading just the calendar keyword. Basically all of our APIs do a `.try_into()` on the locale before running, we just do this with a slightly different type. AnyCalendarKind is just "the relevant subset of the datalocale", to me.

Comment (sffc) [reply ffi/diplomat/src/calendar.rs:58]: I think the other issue here is that failing to turn a `Locale` into a `AnyCalendarKind` shoudn't fail to create an AnyCalendar; we're supposed to fall back to the locale's default. That is all code that should ideally live in Rust, not in the FFI wrapper layer or in the C++/JavaScript calling code.

I guess I would expect a function like `AnyCalendar::try_new_from_locale_unstable(&provider, &locale)`

File: ffi/diplomat/src/calendar.rs
Comment (sffc) [reply ffi/diplomat/src/calendar.rs:82]: I guess the other option would be to make the `second` argument an `f64`... but then we get into ryu stuff pretty quickly (although we could likely use the fast path to avoid ryu).

File: ffi/diplomat/wasm/wasm-demo/src/ts/date-time.ts
Comment (sffc) [inline ffi/diplomat/wasm/wasm-demo/src/ts/date-time.ts:159]: Thought: This solution doesn't appear to be any worse than other solutions. Here is how you would do it in Temporal:

https://stackoverflow.com/a/73298208/1407170

File: ffi/diplomat/wasm/wasm-demo/src/ts/app.ts
Comment (sffc) [inline ffi/diplomat/wasm/wasm-demo/src/ts/app.ts:8]: Issue: I think TypeScript still won't like these boostrap includes... but this PR is still a net improvement.

**Review by sffc — APPROVED** [review]

File: ffi/diplomat/wasm/wasm-demo/src/ts/date-time.ts
Comment (Manishearth) [reply ffi/diplomat/wasm/wasm-demo/src/ts/date-time.ts:159]: Yeah the input element's `valueAsDate` probably should interop with Temporal, eventually.

(Perhaps valueAsTemporalDate or valueAsDateTime)

**Pull request merged**
